### PR TITLE
Header in CSV file enhancement

### PR DESCRIPTION
**Pull request by chripard**
Please review.

### CODE REVIEW

File: src/main/java/org/repodriller/persistence/csv/CSVFile.java [hunk 1]
```diff
@@ -18,14 +15,18 @@
...
 
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringEscapeUtils;
 import org.repodriller.persistence.PersistenceMechanism;
 
 public class CSVFile implements PersistenceMechanism {
 
 	private PrintStream ps;
-
+	private String[] header = new String[0];
```
Comment (mauricioaniche) [inline src/main/java/org/repodriller/persistence/csv/CSVFile.java]: What happens if the user doesn't pass a header? Are you sure it works? Maybe an automated test to proof it? :)

File: src/main/java/org/repodriller/persistence/csv/CSVFile.java [hunk 2]
```diff
@@ -44,10 +45,34 @@ public CSVFile(String path, String name) {
...
+			ps.println(headers);		
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+	}
+	
 	@Override
-	public synchronized void write(Object... line) {
+	public synchronized void write(Object... line) throws CSVFileFormatException {		
+		if (header.length == 0) header = ArrayUtils.clone(new String[line.length]);
```
Comment (mauricioaniche) [inline src/main/java/org/repodriller/persistence/csv/CSVFile.java]: why do you need this if? It means that every time we write, we need to allocate a new array. I think it's simpler to do `if(header!=null && header.length != line.length)`.

Then, you make `header=null` in the constructors that do not receive the header

File: src/test/java/org/repodriller/persistence/csv/CSVFileTest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+	private String pathToRepo7;
+
+	@Before
+	public void setUp() {
+		this.pathToRepo7 = this.getClass().getResource("/").getPath() + "../../test-repos/git-7";
+	}
+	
+	@Test(expected = CSVFileFormatException.class)
+		public void createCSV() {
```
Comment (mauricioaniche) [inline src/test/java/org/repodriller/persistence/csv/CSVFileTest.java]: the name of this one can be `valuesDoNotMatchHeaders`.

File: src/test/java/org/repodriller/persistence/csv/CSVFileTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		TestCSVisitor visitor = new TestCSVisitor();
+		CSVFile file = new CSVFile("C:/Users/dev/Desktop","test.csv", new String [] {"Hash"});
+		
+		new RepositoryMining()
+			.in(GitRepository.singleProject(pathToRepo7))
+			.through(Commits.all())
+			.process(visitor, file)
+			.mine();
+		}
+
```
Comment (mauricioaniche) [inline src/test/java/org/repodriller/persistence/csv/CSVFileTest.java]: you should also test:
- One single column working
- Many columns working
- One single column failing
- Many columns failing

To test it, you will have to really write the file somewhere and open it to see if it has the structure you want. I suggest you to use `String path = System.getProperty("java.io.tmpdir") + "file.csv")` to get the path to a temporary directory. Then, every time the test runs (`@Before`), you delete this file. 

Does that make sense to you?

**Comment by mauricioaniche** [comment]
@chripard I just created some tests to `CSFile` as the other PR had a bug (if wasn't handling 'null' parameters). Pull it and then make sure to add tests about this new feature in this class!

**Comment by mauricioaniche** [comment]
Hi Christina,

There are conflicts and `<<<<` in the code! This cannot be merged :/

**Comment by chripard** [comment]
Shall i close this and pull a new one?

**Comment by mauricioaniche** [comment]
I think so. The way it is, it's easier for you to copy your changes to a safe place, and start it over again from the master branch.

**Pull request closed without merging**